Refactor ConstantAlpha to allow running with turbulence model

ewquon opened the pull request:
Previous approach:
* Calculate shear stress (tau), which in general is strain rate multiplied by 2*(mu + mu_t) -- here mu and mu_t refer to the molecular and turbulent viscosities
* When calculating the diffusion contribution, multiply by rho/rho0 if using ConstantAlpha -- here, rho is the instantaneous local value, _averaged from cell centers to faces_
* The issue arises because mu_t also ends up being scaled by this density ratio

New approach
* Calculate shear stress with appropriate mu_eff, based on whether we're using the Constant or ConstantAlpha solver path -- here, tau is scaled by the _cell-centered or face-averaged_ value of rho, depending on the stress quantity
* Diffusion contribution calculations are agnostic to the molec_diff_type
* **Because the density used to scale the diffusion contribution is different, the gold data for regtests with ConstantAlpha will change**

Tests based on DensityCurrent
* ConstantAlpha with `les_type=none`: `dynamicViscosity=75`, `rho0_trans=1` (default) vs `dynamicViscosity=150`, `rho0_trans=2` -- verified results were identical
* ConstantAlpha with `les_type=Smagorinsky`: `dynamicViscosity=75`, `rho0_trans=1` (default) vs `dynamicViscosity=150`, `rho0_trans=2` -- verified results were identical
* very small differences observed with and without Smagorinsky in this case

ewquon committed:
Enable ConstantAlpha with LES/PBL

ewquon committed:
Throw error if Smagorinsky selected w/o setting Cs

ewquon committed:
Fix bug in incompressible path

ewquon committed:
Cleanup: variable names for clarity and consistency, update comments

ewquon committed:
Handle ConstantAlpha case when calling ComputeStress*

- ComputeStress*() now have cell_data as an argument
- cell_data is set if l_use_constAlpha, otherwise a nullptr is passed
- mu_eff is now converted to alpha prior to being passed to
  ComputeStress*()

ewquon committed:
Apply the same changes for ConstantAlpha as ERF_slow_rhs_pre

ewquon committed:
No need to scale by rho here in the ConstantAlpha solver path

For generality, rho scaling is applied when tau is calculated. This
enables ConstantAlpha to be used with a turbulence model.

AMLattanzi committed:
Merge branch 'development' into constalpha_with_turb

AMLattanzi committed:
Merge branch 'development' into constalpha_with_turb

ewquon commented:
Density current prob (coarse dx=dz=100 m) at 900 s, before/after this PR
![compare_before_after_PR1391](https://github.com/erf-model/ERF/assets/18267059/e3e922cb-4ff0-431c-bf21-4c7be945425a)

ewquon committed:
Update golds after ConstantAlpha refactor

ewquon committed:
Make interpolation of rhoAlpha consistent with mu_turb

ewquon committed:
Calculate rhoalpha for ghost cells in xyz

ewquon committed:
Revert "Update golds after ConstantAlpha refactor"

This reverts commit 71f1c311a71a39843cc8f09bb96e02126ebd4b7d.

AMLattanzi commented on Source/Diffusion/ComputeStress_N.cpp:
```diff
@@ -20,32 +21,60 @@ using namespace amrex;
+        },
+        [=] AMREX_GPU_DEVICE (int i, int j, int k) noexcept {
+            tau23(i,j,k) *= -cell_data(i, j, k, Rho_comp) * mu_eff;
+        });
```
For the off-diagonal stresses, shouldn't the **cell_data** be averaged to the location of the strain (this would hold for other places in the code as well)?
![image](https://github.com/erf-model/ERF/assets/103702284/78fce89e-c942-42ea-bd32-f13f16c6c691)

ewquon replied:
Yes, you're right. I have a fix coming.

AMLattanzi reviewed (changes requested):
Fairly certain the cell_data must be averaged to the strain locations (as is done in **DiffusionSrcForState_N/T.cpp**. Once that is completed, I am good with this PR.

ewquon committed:
New gold data after fixing indexing and interpolation

ewquon commented:
@AMLattanzi, [23d55a5](https://github.com/erf-model/ERF/pull/1391/commits/23d55a50e6f9ca3ae393f319f1eab6055a027799) fixes the segfault that showed up in the CI and [1834373](https://github.com/erf-model/ERF/pull/1391/commits/18343734fc5d21f155740303fd5c8ab04ee75357) does the averaging that's needed.

ewquon commented:
Actually hold on, I only updated the terrain solver path.

ewquon committed:
Average rho to faces, following mu_turb; update gold data

ewquon commented:
Ok, @AMLattanzi, this latest commit makes the rest of the changes. I've verified that the rho averaging for the ConstantAlpha path matches the original mu_turb averaging.

ewquon committed:
Removed fixed_mri_dt_ratio

Otherwise ref solution crashes on step 938 (t=234.25). Orig fixed dt ratio was
4; allowing a variable dt ratio results in a constant dt ratio of 6.
